### PR TITLE
Filter out TWAP orders from not allowed apps

### PR DESCRIPTION
**Pull request by hectorgomezv**
## Summary
This PR implements `TwapOrderHelper.isAppAllowed`, which prevents the decoding of TWAP orders emitted by non-allowed apps.

## Changes
- Add `TwapOrderHelper.isAppAllowed`.
- Add test cases verifying the non-decoding of these orders.

### CODE REVIEW

File: src/routes/transactions/transactions-view.service.ts [hunk 1]
```diff
@@ -172,9 +172,17 @@ export class TransactionsViewService {
...
     });
 
-    const [{ fullAppData }, buyToken, sellToken] = await Promise.all([
-      // Decode hash of `appData`
-      this.swapsRepository.getFullAppData(args.chainId, twapStruct.appData),
+    // Decode hash of `appData`
+    const fullAppData = await this.swapsRepository.getFullAppData(
+      args.chainId,
+      twapStruct.appData,
+    );
```
Comment (hectorgomezv) [inline src/routes/transactions/transactions-view.service.ts:179]: I've decupled this block from the tokens retrieval, so we can fail-fast if the app is not allowed.

File: src/routes/transactions/helpers/twap-order.helper.ts [hunk 2]
```diff
@@ -167,6 +177,20 @@ export class TwapOrderHelper {
     });
   }
 
+  /**
+   * Checks if the app associated contained in fullAppData is allowed.
+   *
+   * @param fullAppData - object to which we should verify the app data with
+   * @returns true if the app is allowed, false otherwise.
+   */
+  public isAppAllowed(fullAppData: FullAppData): boolean {
```
Comment (iamacook) [inline src/routes/transactions/helpers/twap-order.helper.ts:186]: This duplicates that in the `SwapOrderHelper`. Maybe we should extract it somehow - what do you think? (I created a `disallow-twap-mapping` branch before going away that might be possible to reference.)

**Review by iamacook — COMMENTED** [review]
We also need to add this to `SwapTransferInfoMapper['mapSwapTransferInfo']` after finding the order by transfer.

File: src/routes/transactions/mappers/common/twap-order.mapper.ts [hunk 3]
```diff
@@ -111,7 +111,10 @@ export class TwapOrderMapper {
       }
     }
 
-    // TODO: Handling of restricted Apps, calling `getToken` directly instead of multiple times in `getOrder` for sellToken and buyToken
+    if (!this.twapOrderHelper.isAppAllowed(fullAppData)) {
+      throw new Error(`Unsupported App: ${fullAppData.fullAppData?.appCode}`);
+    }
```
[the remark this answers is not in the document]
Comment (hectorgomezv) [reply src/routes/transactions/mappers/common/twap-order.mapper.ts:126]: Added in [6c540a0](https://github.com/safe-global/safe-client-gateway/pull/1698/commits/6c540a0e504a2a4784c74187530848723f076e26)

File: src/routes/transactions/mappers/common/twap-order.mapper.ts [hunk 4]
```diff
@@ -101,6 +101,16 @@ export class TwapOrderMapper {
...
       try {
         const order = await this.swapsRepository.getOrder(chainId, orderUid);
+        const partFullAppData = await this.swapsRepository.getFullAppData(
+          chainId,
+          part.appData,
+        );
+        if (!this.twapOrderHelper.isAppAllowed(order)) {
+          throw new Error(
+            `Unsupported App: ${partFullAppData.fullAppData?.appCode}`,
+          );
```
Comment (iamacook) [inline src/routes/transactions/mappers/common/twap-order.mapper.ts:111]:
```suggestion
        if (!this.twapOrderHelper.isAppAllowed(order)) {
          throw new Error(
            `Unsupported App: ${order.fullAppData?.appCode}`,
          );
```

File: src/routes/transactions/mappers/common/twap-order.mapper.ts [hunk 5]
```diff
@@ -85,7 +85,7 @@ export class TwapOrderMapper {
         : twapParts
       : [];
 
-    const { fullAppData } = await this.swapsRepository.getFullAppData(
+    const fullAppData = await this.swapsRepository.getFullAppData(
```
Comment (iamacook) [inline src/routes/transactions/mappers/common/twap-order.mapper.ts:88]: We can move this below the checking of each part to reude request.

**Review by iamacook — COMMENTED** [review]
I think we also need check the `appData` of [the "found" order](https://github.com/safe-global/safe-client-gateway/blob/restrict-twap-apps/src/routes/transactions/mappers/transfers/swap-transfer-info.mapper.ts#L56) when mapping swap transfers as well. We will otherwise map all swap transfers.

**Review by iamacook — APPROVED** [review]

**Pull request merged**